### ISSUE TITLE
Register application to handle ss:// links

### ISSUE THREAD

**Issue by jab** [shadowsocks/shadowsocks-windows#1146]
It seems like it would make it easier to add a server if shadowsocks-windows registered itself to handle ss:// URLs. Then you could simply click an ss:// link to add a server to shadowsocks-windows.

It looks like it's just a matter of editing the registry:
- [Installing and Registering Protocol Handlers (MSDN)](https://msdn.microsoft.com/en-us/library/windows/desktop/bb266527(v=vs.85).aspx)
- [Registering an Application to a URI Scheme (MSDN)](https://msdn.microsoft.com/en-us/library/aa767914(v=vs.85).aspx)

Interested?

**Comment by wongsyrone** [comment]
Good point but Shadowsocks meant to be portable. I don't want to pollute user's registry.

**Comment by jab** [comment]
Thanks for the quick response. Is there anything that could lead you to reconsider, say if enough users upvote this issue? All the users I know would much prefer to be able to click an ss:// link to add a server, and wouldn't care if shadowsocks-windows had to modify the registry to do it.

The broader context here is that based on user testing, there are currently a few usability improvements we could make to shadowsocks-windows that would make it much more efficient for new users to go from zero to shadowsocks-windows installed and proxying through a server that has been shared with them. This is one of those improvements. I'll open separate issues for others.

**Comment by jab** [comment]
Also meant to say, for any issue I open including this one, I would be happy to try to work up a pull request so all you have to do is click merge.

**Comment by hellofwy** [comment]
What if there are multiple ss-win?
May provide a separate installer version to add those features relating to the registry.

**Comment by celeron533** [comment]
@jab , perhaps QR code is a workaround.

![image](https://cloud.githubusercontent.com/assets/3608762/26252982/616d41ae-3ce4-11e7-81fa-de4dc160021d.png)

![image](https://cloud.githubusercontent.com/assets/3608762/26252987/64796c06-3ce4-11e7-940b-99c75831c1a7.png)

![image](https://cloud.githubusercontent.com/assets/3608762/26252996/698242fe-3ce4-11e7-86e2-85fe2bc60469.png)

**Comment by jab** [comment]
Thanks @celeron533, but user testing showed that adding via QR code was a very difficult and unusual UX for new users on a desktop OS.

@hellofwy wrote:
> What if there are multiple ss-win? May provide a separate installer version to add those features relating to the registry.

If making an installer version is the only way to get this feature landed, let's do it.

But I'm guessing we could land this even sooner and more easily if we made ss-win register itself as the ss:// handler on startup. It solves the problem for users who only have one copy of ss-win, which is 100% of new users.

For users who have multiple copies of ss-win, things are no worse than they are now, since they currently can't click an ss:// link to add a server to any of their copies. If we added this, they could click an ss:// link to add a server to their most recently run copy.

Note that a confirmation dialog could be shown for this, similar to the one I mocked up in #1147 for adding via clipboard:

![mock](https://cloud.githubusercontent.com/assets/64992/26253944/fc4b71a8-3c82-11e7-9794-b1a991599e47.png)

The "never ask again" option could unregister the application as the ss:// handler, giving users the option to opt out, in case this is not what they want.
